### DISCUSSION TITLE
Extracting hashes from encrypted archives

### DISCUSSION THREAD

**Discussion by img507** [miurahr/py7zr#599] [Q&A]
Hello Hiroshi Miura, I would like to ask you to help me with one project. Now I am trying to finish the script for extracting password hashes from 7zip archives, I do it as a CTF tool. This is an old project of another person, which was written for python 2, but now it is no longer relevant.
Here is the link
https://github.com/truongkma/ctf-tools/blob/master/John/run/7z2john.py
You wrote a very good library for working with 7zip archives, you know their structure better. This script can extract passwords only from archives that have an encrypted header and only one file inside, even if the file is in a folder, the script does not work quite correctly. It cannot process archives with unencrypted headers and several files inside. Unfortunately, there is also no support for all compression algorithms. I slightly edited this script for python 3, also completed the ability to output the hash in the format for hashcat (not completely), nothing else has changed dramatically.  I know there is already a script from philsmd that he wrote in perl, but I want to make it in python. I really ask you to help me finish this script, I will attach my file to this message, later I want to post it on github, I will definitely indicate you as a co-author
My file, I uploaded it to pastebin
https://pastebin.com/Y4TaD7Hm

**Comment by miurahr** [comment]
You can find a documentation of 7zip file structure at https://py7zr.readthedocs.io/en/latest/archive_format.html
You can also read  `py7zr/archiveinfo.py`  which represent 7zip header.

It may be good start point that `py7zr/py7zr.py:SevenZipFile` class ` _real_get_contents(self, password)` method that is like
```
       if not self._check_7zfile(self.fp):
            raise Bad7zFile("not a 7z file")
        self.sig_header = SignatureHeader.retrieve(self.fp)
        self.afterheader: int = self.fp.tell()
        self.fp.seek(self.sig_header.nextheaderofs, os.SEEK_CUR)
        buffer = io.BytesIO(self.fp.read(self.sig_header.nextheadersize))
        if self.sig_header.nextheadercrc != calculate_crc32(buffer.getvalue()):
            raise Bad7zFile("invalid header data")
        header = Header.retrieve(self.fp, buffer, self.afterheader, password)
```
You can observe `header` object that contains information which you want.

**Reply by img507** [reply]
Hello Hiroshi Miura, thanks for the answer, I have one idea, you can add my file that reads and extracts hashes only from archives with encrypted headers to your library, add the ability to read archives (with an unencrypted header and all the files inside it, as well as all the compression algorithms that you have already successfully implemented) using the archiveinfo.py file, it seems that it is responsible for this (correct me if I'm wrong) And output the resulting hash in the format for hashcat, in this case we will be able to make a full-fledged ctf tool for extracting hashes from all 7z archives, regardless of the header encryption and the number of files inside. I think this is a great option. I really ask you to help, Hiroshi Miura. I really wanted such a project to be implemented

**Reply by miurahr** [reply]
I don't do it. You can use files in your project from py7zr library.

**Reply by miurahr** [reply]
There is `crc32` field in `py7zr.FileInfo` object.  It is returns as `List[FileInfo]` by `SevenZipFile.list` method.   
You may be able to write like as follows:
 
```python
    with SevenZipFile("foo.7z", "r") as szf:
        for info in szf.info():
            sys.stdout.write("%d".format(info.crc32))
```
